### PR TITLE
Add CheckReturnValue annotation

### PR DESCRIPTION
**Pull request by rillig**

### CODE REVIEW

File: common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * Specifies that the return value of the method should be used.
+ * <p>
+ * Typical use cases are:
+ * <ul>
+ *   <li>pure or side-effect-free methods
+ *   <li>methods that use the return value for reporting success or error
+ *   <li>methods that return an unfinished computation,
+ *       such as {@link Collection#stream()} or builders
```
Comment (amaembo) [inline common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java:17]: This is basically the same as "pure or side-effect-free" methods, so I don't see the difference. I would actually concentrate on side-effect methods, like

> Specifies that the return value of the impure method should be used. For pure method (annotated with `@Contract(pure = true)`), it's implied that the resulting value is important, so this annotation is unnecessary. However, sometimes it's important for impure method as well. For example, `InputStream#read(byte[])` returns the number of bytes actually stored into byte array. Without checking the return value, it's impossible to say how many bytes were actually read.

WDYT?

File: common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package org.jetbrains.annotations;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Collection;
+
+/**
+ * Specifies that the return value of the method should be used.
```
Comment (chashnikov) [inline common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java:10]: I think 'should' is a too weak word here, I would use 'must'.

File: common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * </ul>
+ * <p>
+ * When used on a type, the annotation applies to all methods that do not return {@code void}.
+ * <p>
+ * When used on a package, the annotation applies to all types of that package.
+ *
+ * @see Contract#pure()
+ */
+@Target({ElementType.METHOD, ElementType.CONSTRUCTOR, ElementType.TYPE, ElementType.PACKAGE})
+@Retention(RetentionPolicy.RUNTIME)
```
Comment (chashnikov) [inline common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java:27]: Why do you use `RUNTIME` retention for this annotation? It's explicitly forbidden in [CONTRIBUTING.md](https://github.com/JetBrains/java-annotations/blob/master/CONTRIBUTING.md#retention-policy).

File: common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+@ApiStatus.Experimental
+public @interface CheckReturnValue {
```
Comment (chashnikov) [inline common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java:29]: And please increase the major version as described in [CONTRIBUTING.md](https://github.com/JetBrains/java-annotations/blob/master/CONTRIBUTING.md).

File: common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java (hunk 1 above)
Comment (rillig) [reply common/src/main/java/org/jetbrains/annotations/CheckReturnValue.java]: Sounds good, thanks for the example.

**Pull request merged**